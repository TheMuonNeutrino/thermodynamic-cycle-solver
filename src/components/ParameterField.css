.parameter-field-value{
    width: 120px !important; 
}
.parameter-field{
    width: fit-content !important;
}