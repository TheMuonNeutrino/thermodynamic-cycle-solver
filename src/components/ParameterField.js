import React, { useCallback, useEffect } from 'react';
import './ParameterField.css';

const ParameterField = ({label,value,readOnly,updateValue,updateKey}) =>{

    var [localValue, setLocalValue] = React.useState(value)

    var updateLocalValue = () => {}
    if (!readOnly){
        updateLocalValue = (value) => {
            if (value === ''){
                setLocalValue('')
            }else{
                const newVal = parseFloat(value)
                if (!isNaN(newVal)){
                    setLocalValue(newVal)
                }
            }
        }
    }

    const onChangeParam = useCallback(() => {
        if (localValue !== '' && localValue !== value){
            var pointChanges = {}
            pointChanges[updateKey] = localValue
            updateValue(pointChanges)
        }
    },[localValue,updateValue,updateKey,value])

    useEffect(()=>{
        setLocalValue(value)
    },[value])

    useEffect(()=>{
        var timeoutRef = setTimeout(()=>{
            onChangeParam()
        },200)
        return(()=>{
            clearTimeout(timeoutRef)
        })
    },[localValue,onChangeParam])

    return(
        <div className='field parameter-field'>
            <label className='parameter-field-label'>{label}</label>
            <input 
                type='text'
                className='parameter-field-value' 
                value={localValue} 
                readOnly={readOnly}
                onBlur={onChangeParam}
                onChange={(e)=>{updateLocalValue(e.target.value)}}
                disabled={readOnly}
            />
        </div>
    )
}

export default ParameterField