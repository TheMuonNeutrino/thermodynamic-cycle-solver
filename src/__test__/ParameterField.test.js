import React from 'react';
import { shallow } from 'enzyme';
import ParameterField from "../components/ParameterField";

describe('ParameterField tests',()=>{
    var value, onValueChange, component
    beforeEach(()=>{
        value = 0
        onValueChange = jest.fn()
        component = shallow(
            <ParameterField 
                label='myTestLabel' 
                value='myTestValue' 
                readOnly={false} 
                updateValue={onValueChange}
                updateKey='myTestUpdateKey'
            />
        )
    })
    it('Should render with label and value',()=>{
        expect(component.getElements()).toMatchSnapshot()
    })
    it('Should call onValueChange when changed',()=>{
        component.find('input').simulate('change',{
            target: {value: '10'}
        })
        expect(onValueChange).toHaveBeenLastCalledWith({myTestUpdateKey: '10'})
    })
})