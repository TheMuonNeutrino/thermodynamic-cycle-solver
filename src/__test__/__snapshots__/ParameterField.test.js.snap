// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParameterField tests Should render with label and value 1`] = `
Array [
  <div
    className="field parameter-field"
  >
    <label
      className="parameter-field-label"
    >
      myTestLabel
    </label>
    <input
      className="parameter-field-value"
      disabled={false}
      onBlur={[Function]}
      onChange={[Function]}
      readOnly={false}
      type="text"
      value="myTestValue"
    />
  </div>,
]
`;
